crates/forge/src/lib.rs

```diff
@@ -72,6 +74,48 @@ struct TestsFromFile {
     relative_path: Utf8PathBuf,
 }
 
+#[must_use]
+pub fn collect_packages(
+    metadata: &Metadata,
+    manifest_path: &Utf8PathBuf,
+    package_name: &Option<String>,
+    workspace: bool,
+) -> Vec<PackageId> {
+    let workspace_members = &metadata.workspace.members;
+
+    let is_virtual_workspace = metadata
+        .packages
+        .iter()
+        .filter(|package| package.manifest_path == metadata.workspace.manifest_path)
+        .collect_vec()
+        .is_empty();
+
+    let package_filter: Box<dyn Fn(&PackageMetadata) -> bool> = {
+        if workspace {
+            Box::new(|_| true)
+        } else if let Some(package_name) = package_name {
+            // selected specific package
+            Box::new(|package| package.name == *package_name)
+        } else if metadata.workspace.manifest_path != *manifest_path {
+            // snforge command ran inside package
+            Box::new(|package| package.manifest_path == *manifest_path)
+        } else {
+            // snforge command ran in root directory
+            Box::new(|package| {
+                package.manifest_path == metadata.workspace.manifest_path || is_virtual_workspace
+            })
+        }
+    };
+
+    metadata
+        .packages
+        .iter()
+        .filter(|package| workspace_members.contains(&package.id))
```

Comment:
Cannot we just filter `workspace_members`? We return ids only anyways

Reply:
We use metedata of packages in the filtering function, so we need additional fields, which aren't present in the `workspace_members` vector.

Reply:
Right, my bad